daemon/cluster/services.go

```diff
@@ -303,35 +300,37 @@ func (c *Cluster) RemoveService(input string) error {
 }
 
 // ServiceLogs collects service logs and writes them back to `config.OutStream`
-func (c *Cluster) ServiceLogs(ctx context.Context, selector *backend.LogSelector, config *backend.ContainerLogsConfig, started chan struct{}) error {
-	c.mu.RLock()
+func (c *Cluster) ServiceLogs(ctx context.Context, selector *backend.LogSelector, config *backend.ContainerLogsConfig) (rerr error) {
+
+	// it's either error here or we will either block on trying to send, or
+	// panic when we try to close. the former case is gonna leave someone
+	// scratching their head for a few minutes before they realize their dumb
+	// mistake
+	if config.Messages == nil {
+		// i'm REALLY tempted to do panic() here because i NEVER get to use panic()
+		return errors.New("can't send messages on nil channel")
+	}
+	// if we return nil, it means we're still using this channel, so don't close
+	// it
 	defer func() {
-		select {
-		case <-started:
-			// if we've started streaming logs, we are no longer holding the
-			// lock and do not have to release it
-			return
-		default:
-			// before we start, though, we're holding this lock and it needs to
-			// be released
-			c.mu.RUnlock()
+		if rerr != nil {
+			close(config.Messages)
```

Comment:
As I commented on `ContainerLogs`, I think this function should return a channel.